htlcswitch/switch.go

```diff
@@ -0,0 +1,399 @@
+//	alice			   bob				   carol
+//	server - <-connection-> - server - - <-connection-> - - - server
+//	 |		   	  |				    |
+//   alice htlc			bob htlc		       carol htlc
+//     switch			switch			         switch
+//	|			 |       \			   |
+//	|			 |        \			   |
+// alice htlc  <-channel->  first bob    second bob <-channel-> carol htlc
+// manager	    	  htlc manager   htlc manager		manager
+type HTLCSwitch struct {
+	service *Service
+	mutex   sync.RWMutex
+
+	// circuits is an structure which is used to forward the settle HTLC
+	// back to the add HTLC initiator.
+	circuits *circuitMap
+
+	// managers is a map of channel output and HTLC manager which manages
+	// this channel.
+	managers map[wire.OutPoint]HTLCManager
+
+	// commands...
+	commands chan *handleRequestCommand
+}
+
+// NewHTLCSwitch creates the HTLCSwitch instance.
+func NewHTLCSwitch() (*HTLCSwitch, error) {
+	circuitMap, err := newCircuitMap()
+	if err != nil {
+		return nil, err
+	}
+
+	return &HTLCSwitch{
+		service:  NewService("HTLC Switch"),
+		circuits: circuitMap,
+		managers: make(map[wire.OutPoint]HTLCManager),
+		commands: make(chan *handleRequestCommand),
+	}, nil
+}
+
+// Forward is used by HTLC managers to propagate the HTLC after it isn't
+// reached its final destination and eligible for forwarding. HTLC are
+// encapsulated in switch request in order to carry additional information.
+func (s *HTLCSwitch) Forward(request *SwitchRequest) error {
+	command := &handleRequestCommand{
+		request: request,
+		err:     make(chan error),
+	}
+
+	select {
+	case s.commands <- command:
+		return <-command.err
+	case <-s.service.Quit:
+		return nil
+	}
+}
+
+// forwardRequest handles incoming forward requests received from htlc managers.
+func (s *HTLCSwitch) forwardRequest(request *SwitchRequest) error {
+	switch request.Type {
+
+	// User sent us new payment request, therefore we trying to find the
+	// HTLC appropriate manager in terms of destination and bandwidth.
+	case UserAddRequest:
+		htlc := request.Htlc.(*lnwire.HTLCAddRequest)
+
+		managers, err := s.getManagersByDest(request.Dest)
+		if err != nil {
+			return err
+		}
+
+		var destination HTLCManager
+		for _, manager := range managers {
+			if manager.Bandwidth() >= htlc.Amount {
+				destination = manager
+				break
+			}
+		}
+
+		if destination == nil {
+			request.Error() <- errors.New("unable to send payment, " +
+				"insufficient bandwidth")
+			return nil
+		}
+
+		log.Debugf("Sending %v to %x", htlc.Amount, request.Dest.String())
+		return destination.HandleRequest(request)
+
+	// HTLC manager forwarded us a new HTLC, therefore we initiate the
+	// payment circuit within our internal state so we can properly forward
+	// the ultimate settle message back latter.
+	case ForwardAddRequest:
+		htlc := request.Htlc.(*lnwire.HTLCAddRequest)
+
+		source, err := s.Get(request.ChannelPoint)
+		if err != nil {
+			return errors.Errorf("unable to find source htlc "+
+				"manager %v", err)
+		}
+
```

Comment:
The _exact_ logging messages which were present prior to this PR should be restored.